### ISSUE TITLE
sounddevice re-assigns stderr, leading to hard-to-track issues

### ISSUE THREAD

**Issue by kmbond** [psychopy/psychopy#2230]
I recently transitioned to psychopy3, and now I’m having issues gathering user input using `input()`, but only after I import the sound module from psychopy. This worked in psychopy 1.9 using `raw_input()`. 

The prompt for the input doesn’t show after I import the sound module (and only the sound module, works fine if I import others), but I can enter a value and it is stored. I don't receive an error message when this happens. 

This behavior even persists if I rename the input function from python 3 (thought that there might be a namespace conflict). **This suggests to me some serious coding issue at a basic level, considering that the behavior of base functions in python are being corrupted.** The script below tests this odd behavior.

```python3
from psychopy import prefs
prefs.general['audioLib'] = ['pygame','pyo']
from builtins import input as new_input #renaming the builtin input function for python to test namespace conflict 
new_input("Testing? ") #this line produces a prompt in the terminal 
from psychopy.sound import Sound
test_input=new_input("Testing? ") #this line will not 
print(test_input) #but the value entered after the line above is run is stored 
```

**Comment by hoechenberger** [comment]
I would speculate that the prompt gets corrupted, but not a “Python base function” per se. In either case, as you figured, assigning the function name to a new name, which points to the exact same function in memory, won’t help. Which version of PsychoPy are you using, specifically (standalone or manually installed via pip or conda; which version exactly) on which platform? In case you’re not using the standalone package: which audio backends are installed? Thanks!

**Comment by kmbond** [comment]
Thanks for the quick response. 

I'm using Pyschopy 3.0.0, installed via conda 4.5.11. The platform is Ubuntu 18.04.1. The pygame audio backend is installed. Let me know if you need more information.

**Comment by hoechenberger** [comment]
Thanks for your feedback and detailed descriptions, @kmbond! I don't have a Linux box available right now, but will try to reproduce this later today on my macOS system. May I ask which kind of terminal emulator and shell you are using?

**Comment by hoechenberger** [comment]
Cannot reproduce on macOS with PsychoPy `master`, Python 3.7.1 installed via `conda-forge`; `pygame` was installed via `pip`.

```
$ conda env export
name: psychopy-test
channels:
  - conda-forge
  - defaults
dependencies:
  - blas=1.0=mkl
  - bzip2=1.0.6=h1de35cc_1002
  - ca-certificates=2018.11.29=ha4d7672_0
  - certifi=2018.11.29=py37_1000
  - configobj=5.0.6=py_0
  - future=0.17.1=py37_1000
  - intel-openmp=2019.1=144
  - json_tricks=3.12.2=py_0
  - libcxx=7.0.0=h2d50403_2
  - libffi=3.2.1=h0a44026_1005
  - libgfortran=3.0.1=h93005f0_2
  - llvm-meta=7.0.0=0
  - mkl=2019.1=144
  - mkl_fft=1.0.10=py37h1de35cc_1
  - mkl_random=1.0.2=py37h1702cab_2
  - ncurses=6.1=h0a44026_1002
  - numpy=1.15.4=py37hacdab7b_0
  - numpy-base=1.15.4=py37h6575580_0
  - openblas=0.3.3=hdc02c5d_1001
  - openssl=1.0.2p=h1de35cc_1002
  - pandas=0.24.0=py37h0a44026_0
  - pip=18.1=py37_1000
  - python=3.7.1=h145921a_1000
  - python-dateutil=2.7.5=py_0
  - pytz=2018.9=py_0
  - readline=7.0=hcfe32e1_1001
  - setuptools=40.6.3=py37_0
  - six=1.12.0=py37_1000
  - sqlite=3.26.0=h1765d9f_1000
  - tk=8.6.9=ha441bb4_1000
  - wheel=0.32.3=py37_0
  - xz=5.2.4=h1de35cc_1001
  - zlib=1.2.11=h1de35cc_1004
prefix: /Users/hoechenberger/miniconda3/envs/psychopy-test
```

How do you run your the test script?

**Comment by peircej** [comment]
I can't replicate either yet (tried windows and mac both using python.org Python). Trying to get this up on an ubuntu box (it's currently running many updates for me). 

In the meantime, my guess is that pygame (or other sound backend, but most likely pygame) is blocking the interactivity to create an event loop. Things you could try:

* progressively comment-out sections of `sound.__init__` to see which lines make the issue go away
* just import pygame and initialize sound to see if that's the problem
* switch to a different way of getting input (e.g. dialog box or visual.Window)

**Comment by peircej** [comment]
I've now checked ubuntu with stock Python and that's fine as well:
![virtualbox_ubuntu64_28_01_2019_13_30_48](https://user-images.githubusercontent.com/380582/51839367-f3df9880-2300-11e9-84ef-046aa9a1190e.png)

**Comment by kmbond** [comment]
I won’t be able to respond in depth until later, but someone on the psychopy forum pointed out a similar issue from a few years ago: https://github.com/spatialaudio/python-sounddevice/issues/61

And a colleague was able to reproduce the error on his system. More details to come...

**Comment by hoechenberger** [comment]
So the fix / workaround could be to simply import readline in sound.py. Seems worth a shot.

**Comment by pytaunay** [comment]
I'm the colleague in question : ] I was notified of this error and found it particularly interesting. I was able to reproduce it with stock python 3.5.2 on Ubuntu 16.04, with psychopy 3.0.2 installed with pip.

Using strace I managed to get to the "why" the output is not produced. The first call to "new_input" invokes a "write" system call to STDERR: 
`
write(2, "Testing? ", 9)
`
After the second import is called, a bunch of open/write calls are performed. However, the one of interest is:
`
write(3, "Testing? ", 9)
`

The system call is done on a _new_ file descriptor that was somehow created by the import command; I have watched it appear under /proc/<pid>/fd/ Now, file descriptor 3 is created with an "open" command. As to why it does not display anything, let's see to what this FD points... strace comes in handy again:
`
open("/dev/null", O_RDWR|O_CLOEXEC)     = 3
`

So, somehow the second import gets to open /dev/null and from then on what should be displayed on the terminal on STDERR gets redirected to /dev/null. As to why this is happening, not quite sure yet.

**Comment by hoechenberger** [comment]
Interesting! Just for testing, could you please uninstall `sounddevice`, set the `audioLib` pref to `['pygame']`, and try again?

```python
from psychopy import prefs
prefs.general['audioLib'] = ['pygame']

from psychopy.sound import Sound
input('foo')
```

**Comment by hoechenberger** [comment]
And, additionally, could you do:
```python
from psychopy.sound.backend_pygame import SoundPygame
input('foo')
```
and report how it behaves?

**Comment by pytaunay** [comment]
These worked correctly. Here is the original one:

```
> python3 test.py
Testing? YES
pygame 1.9.4
Hello from the pygame community. https://www.pygame.org/contribute.html
Testing? NO
NO
2.4341 	WARNING 	sounddevice audio lib was requested but not loaded: sounddevice not working
2.4426 	WARNING 	pyo audio lib was requested but not loaded: ImportError("No module named 'pyo'",)
```

And the test with import SoundPygame:
```
python3 test2.py
pygame 1.9.4
Hello from the pygame community. https://www.pygame.org/contribute.html
fooYES
0.2227 	WARNING 	sounddevice audio lib was requested but not loaded: sounddevice not working
0.2322 	WARNING 	pyo audio lib was requested but not loaded: ImportError("No module named 'pyo'",)
```

sounddevice is the culprit. Looking at that module source code, _stderr is redefined to go to /dev/null_!
Here is the offending content of sounddevice.py:

```
2670 def _ignore_stderr():
2671     """Try to forward PortAudio messages from stderr to /dev/null."""
2672     try:
2673         stdio = _ffi.dlopen(None)
2674         devnull = stdio.fopen(_os.devnull.encode(), b'w')
2675     except (OSError, AttributeError):
2676         return
2677     try:
2678         stdio.stderr = devnull
2679     except _ffi.error:
2680         try:
2681             stdio.__stderrp = devnull
2682         except _ffi.error:
2683             stdio.fclose(devnull)
2684 
2685 
2686 _ignore_stderr()
2687 _atexit.register(_exit_handler)
2688 _initialize()
```
I added a simple print statement on line 2672 (print("I AM IGNORING STDERR")) to make sure this is called. Sure enough:
```
> python3 test.py 
Testing? YES
I AM IGNORING STDERR
NO
NO
```
Commenting out line 2686 solves the problem and even more info is shown on the terminal:
```
>python3 test.py 
Testing? YES
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for 4294967295, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for 4294967295, skipping unlock
Testing? NO
NO
```

Everything makes sense now... sounddevice redefines stderr as /dev/null. Anything using stderr thereafter is sent to the black hole. 

See pull request https://github.com/spatialaudio/python-sounddevice/pull/12

**Comment by pytaunay** [comment]
So it is pretty much what this link found out: https://github.com/spatialaudio/python-sounddevice/issues/61

Now, as to why the prompt goes on STDERR, we have to look at the Python C implementation (https://github.com/python/cpython):
1. The built-in function ```input``` is implemented as ```builtin_input_impl``` in Python/bltinmodule.c
2. There is a call to PyOS_Readline to write the prompt and read the content: `s = PyOS_Readline(stdin, stdout, promptstr);`
3. This function is defined in Parser/myreadline.c It then calls ```PyOS_StdioReadline```
4. And inside ```PyOS_StdioReadline```  (line 118 for readline.c in the Python 3.5 kernel)
```    fflush(sys_stdout);
    if (prompt)
        fprintf(stderr, "%s", prompt);
    fflush(stderr);
```

**Comment by hoechenberger** [comment]
Their approach breaks AT LEAST one more portion of PsychoPy, as reported here: 
https://github.com/spatialaudio/python-sounddevice/issues/37

**Comment by hoechenberger** [comment]
@kmbond @pytaunay So an intermediate solution for you could be to simply uninstall `sounddevice`, and (or??) set the `audioLib` prefs to use `pygame` only.

**Comment by pytaunay** [comment]
@hoechenberger done; Created a new issue for the sounddevice team.
